### PR TITLE
node category merging improvements

### LINKED ISSUE #2348: node category merging improvements
This issue is about to enhance to points in current category merging algorithm.

### Current behaviour:
A category can be defined either by an object or a string. 
All nodes on the same level are grouped by category according to the categories "label" property or the atomic string value, where the first node defining the category defines its configuration.

### Expected behaviour
1) The first node with a "non-string" category should define the category's configuration.
2) the category config object should introduce a new property "id", which, if defined, serves as the "collection key" instead of the label property

### PR DESCRIPTION
**Pull request by JohannesDoberer**
Example node config:
```
...navigation:{
nodes: () => [
      {
        pathSegment: 'home',
        label: 'Home',
        globalNav:true,
        children: [
          {
            category:'1',
            pathSegment: 'luigi',
            label: 'luigi',
            viewUrl: '/microfrontend.html'
          },
          {
            pathSegment: 'amfe',
            label: 'a mfe',
            viewUrl: '/modal.html',
            category:'1'
          },
          {
            pathSegment: 'amfe',
            label: 'a mfe',
            viewUrl: '/microfrontend.html',
            category:{ label:'luigi'}
          },
          {
            category:'luigi',
            pathSegment: 'luigi',
            label: 'luigi',
            viewUrl: '/microfrontend.html'
          },
          {
            category:{id:'1', label:'test', collapsible: true},
            pathSegment: 'luigi2',
            label: 'luigi2  ',
            viewUrl: '/microfrontend.html'
          },
        ]
      },
    ],
}
```
Fixes #2348

### CODE REVIEW

**Comment by stanleychh** [comment]
@JohannesDoberer, there is a conflicting file. Please take a look. Thanks

File: core/src/utilities/helpers/navigation-helpers.js [hunk 1]
```diff
@@ -141,13 +150,27 @@ class NavigationHelpersClass {
...
       if (!node.hideFromNav) {
         arr.push(node);
       }
     });
+
+    Object.keys(result).forEach(category => {
+      if (result[category].metaInfo && result[category].metaInfo.id) {
+        result = renameObjKey(result, result[category].metaInfo.id, result[category].metaInfo.label);
+      }
+    });
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:172]:
```suggestion
    Object.keys(result).forEach(category => {
      const metaInfo = result[category].metaInfo;
      if (metaInfo && metaInfo.id) {
         result[metaInfo.label] = result[metaInfo.id];
         delete result[metaInfo.id];
      }
    });
```

Comment (stanleychh) [reply core/src/utilities/helpers/navigation-helpers.js:172]: It seems we can do it this way, then remove `clone` and `renameObjKey` functions on the top.
Please double-check the logic and make sure the application works as expected.

**Review by stanleychh — COMMENTED** [review]
Left a small suggest.

File: core/src/utilities/helpers/navigation-helpers.js (hunk 1 above)
Comment (JohannesDoberer) [reply core/src/utilities/helpers/navigation-helpers.js:172]: Works 👍 . I would keep the id in metaInfo.

File: core/src/utilities/helpers/navigation-helpers.js [hunk 2]
```diff
@@ -141,13 +150,27 @@ class NavigationHelpersClass {
       if (!arr.metaInfo) {
         arr.metaInfo = metaInfo;
       }
+      if (!arr.metaInfo.collapsible && metaInfo.collapsible) {
+        arr.metaInfo.collapsible = metaInfo.collapsible;
+      }
+      if (category && typeof category === 'object' && category.id && category.label) {
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:156]: Please use our existing `isObject` function. https://github.com/SAP/luigi/blob/0a85092eb4985f416bdbddb952c1e19ce25fc9dc/core/src/utilities/helpers/generic-helpers.js#L35

File: core/src/utilities/helpers/navigation-helpers.js [hunk 3]
```diff
@@ -114,8 +115,8 @@ class NavigationHelpersClass {
       let key;
       let metaInfo;
       const category = node[property];
-      if (category && typeof category === 'object') {
-        key = category.label;
+      if (category && GenericHelpers.isObject(category)) {
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:118]:
```suggestion
      if (GenericHelpers.isObject(category)) {
```
`isObject` function already checked if item exists

File: core/src/utilities/helpers/navigation-helpers.js [hunk 4]
```diff
@@ -141,13 +142,30 @@ class NavigationHelpersClass {
...
       if (!arr.metaInfo.categoryUid && key && arr.metaInfo.collapsible) {
         arr.metaInfo.categoryUid = node.parent ? this.getNodePath(node.parent) + ':' + key : key;
       }
       if (!node.hideFromNav) {
         arr.push(node);
       }
     });
+
+    Object.keys(result).forEach(category => {
+      debugger;
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:161]:
```suggestion
```

File: core/src/utilities/helpers/navigation-helpers.js [hunk 5]
```diff
@@ -141,13 +142,30 @@ class NavigationHelpersClass {
       if (!arr.metaInfo) {
         arr.metaInfo = metaInfo;
       }
+      if (!arr.metaInfo.collapsible && metaInfo.collapsible) {
+        arr.metaInfo.collapsible = metaInfo.collapsible;
+      }
+      if (category && GenericHelpers.isObject(category) && category.id && category.label) {
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:148]:
```suggestion
      if (GenericHelpers.isObject(category) && category.id && category.label) {
```

File: core/src/utilities/helpers/navigation-helpers.js [hunk 6]
```diff
@@ -141,13 +142,30 @@ class NavigationHelpersClass {
...
+        arr.metaInfo.label = category.label;
+        arr.metaInfo.id = category.id;
```
Comment (stanleychh) [inline core/src/utilities/helpers/navigation-helpers.js:150]:
```suggestion
        arr.metaInfo = { ...arr.metaInfo, { label: category.label, id: category.id } }
```
Use ES6 spread operator is nicer. ✌️

**Review by stanleychh — COMMENTED** [review]
Few more minor suggestions

File: core/src/utilities/helpers/navigation-helpers.js (hunk 6 above)
Comment (stanleychh) [reply core/src/utilities/helpers/navigation-helpers.js:150]: @JohannesDoberer , adjusted the syntax of suggestion. Please check again. Thanks

**Review by stanleychh — APPROVED** [review]

**Pull request merged**